rpc/public-credentials/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2022 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+pub use public_credentials_runtime_api::PublicCredentialsApi as PublicCredentialsRuntimeApi;
+
+use std::sync::Arc;
+
+use codec::Codec;
+use jsonrpsee::{
+	core::{async_trait, RpcResult},
+	proc_macros::rpc,
+	types::error::{CallError, ErrorObject},
+};
+
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+
+#[rpc(client, server)]
+pub trait PublicCredentialsApi<BlockHash, OuterSubjectId, OuterCredentialId, OuterCredentialEntry> {
+	/// Return a credential that matches the provided root hash and issued to
+	/// the provided subject, if found.
+	#[method(name = "get_credential")]
+	fn get_credential(
+		&self,
+		subject: OuterSubjectId,
+		credential_id: OuterCredentialId,
+		at: Option<BlockHash>,
+	) -> RpcResult<Option<OuterCredentialEntry>>;
+
+	/// Return all the credentials issued to the provided subject.
+	/// The result is a vector of (credential root hash, credential entry).
+	#[method(name = "get_credentials")]
+	fn get_credentials(
+		&self,
+		subject: OuterSubjectId,
+		at: Option<BlockHash>,
+	) -> RpcResult<Vec<(OuterCredentialId, OuterCredentialEntry)>>;
+}
+
+pub struct PublicCredentialsQuery<
+	Client,
+	Block,
+	OuterSubjectId,
+	SubjectId,
+	OuterCredentialId,
+	CredentialId,
+	OuterCredentialEntry,
+	CredentialEntry,
+> {
+	client: Arc<Client>,
+	_marker: std::marker::PhantomData<(
+		Block,
+		OuterSubjectId,
+		SubjectId,
+		OuterCredentialId,
+		CredentialId,
+		OuterCredentialEntry,
+		CredentialEntry,
+	)>,
+}
+
+impl<
+		Client,
+		Block,
+		OuterSubjectId,
+		SubjectId,
+		OuterCredentialId,
+		CredentialId,
+		OuterCredentialEntry,
+		CredentialEntry,
+	>
+	PublicCredentialsQuery<
+		Client,
+		Block,
+		OuterSubjectId,
+		SubjectId,
+		OuterCredentialId,
+		CredentialId,
+		OuterCredentialEntry,
+		CredentialEntry,
+	>
+{
+	pub fn new(client: Arc<Client>) -> Self {
+		Self {
+			client,
+			_marker: Default::default(),
```

Comment:
Yeah exactly. As this is code that we mostly just update whenever some dependencies change, I think it would be good to make the compilation fail is something else is added, so that we are aware of the changes, as we also start to make use of RPC features now.